FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/viewmodel/RefreshPrivateKeysFromEkmViewModel.kt

```diff
@@ -0,0 +1,233 @@
+      val retryAttempts = 6
+      val idToken = GeneralUtil.getGoogleIdToken(
+        context = context,
+        maxRetryAttemptCount = retryAttempts
+      )
+
+      val ekmPrivateResult = repository.getPrivateKeysViaEkm(
+        context = context,
+        ekmUrl = requireNotNull(activeAccount.clientConfiguration?.keyManagerUrl),
+        idToken = idToken
+      )
+
+      if (ekmPrivateResult.status != Result.Status.SUCCESS) {
+        return@withContext when (ekmPrivateResult.status) {
+          Result.Status.EXCEPTION -> {
+            Result.exception(
+              //Thrown network errors can be ignored but not other types of errors
+              GeneralUtil.preProcessException(
+                context = context,
+                causedException = requireNotNull(ekmPrivateResult.exception)
+              )
+            )
+          }
+
+          Result.Status.ERROR -> Result.exception(ApiException(ekmPrivateResult.data?.apiError))
+
+          else -> throw IllegalStateException(
+            "Unsupported status = ${ekmPrivateResult.status} at this step"
+          )
+        }
+      }
+
+      if (ekmPrivateResult.data?.privateKeys?.isEmpty() == true) {
+        return@withContext Result.success(true)
+      }
+
+      val pgpKeyDetailsList = mutableListOf<PgpKeyDetails>()
+      ekmPrivateResult.data?.privateKeys?.forEach { key ->
+        val parsedList = PgpKey.parsePrivateKeys(requireNotNull(key.decryptedPrivateKey))
+          .map {
+            it.copy(
+              passphraseType = KeyEntity.PassphraseType.RAM,
+              importSourceType = KeyImportDetails.SourceType.EKM,
+              tempPassphrase = null
+            )
+          }
+
+        if (parsedList.isEmpty()) {
+          //ask Tom should we do any things here
+          throw IllegalStateException(context.getString(R.string.could_not_parse_one_of_ekm_key))
+        } else {
+          //ask Tom should we check isFullyDecrypted
+
+          //check that all keys were fully decrypted when we fetched them.
+          // If any is encrypted at all, that's an unexpected error, we should throw an exception.
+          parsedList.forEach {
+            if (!it.isFullyDecrypted) {
+              throw IllegalStateException(
+                context.getString(
+                  R.string.found_not_fully_decrypted_key_ask_admin,
+                  it.fingerprint
+                )
+              )
+            }
+          }
+          pgpKeyDetailsList.addAll(parsedList)
+        }
+      }
+
+      handleParsedKeys(activeAccount, pgpKeyDetailsList)
+
+      Result.success(true)
+    }
+
+  private suspend fun handleParsedKeys(
+    activeAccount: AccountEntity,
+    fetchedPgpKeyDetailsList: MutableList<PgpKeyDetails>
+  ) = withContext(Dispatchers.IO) {
+    val context: Context = getApplication()
+    val keysStorage = KeysStorageImpl.getInstance(context)
+    val existingKeyEntities = keysStorage.getRawKeys()
```

Comment:
Are these keys filtered just for this account? (should not return keys of other accounts set up on the app)

Reply:
`KeysStorageImpl` manages keys for an active account only.(starting from #923)